File: PiotBingTST.java
/**
 *  This is a tryout with ternary search tree.
 *  It doesn't solve the problem correctly.
 *
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class PiotBingTST {

	public static class TST<Value> {

		private static class Node {
			private char c;
			private Node left, mid, right;
			private int val;
		}

		private Node root;

		public TST() {
		}

		public boolean contains(final String key) {
			return get(key) != 0;
		}

		private Node get(final Node x, final String key, final int d) {
			if (x == null) {
				return null;
			}
			final char c = key.charAt(d);
			if (c < x.c) {
				return get(x.left, key, d);
			} else if (c > x.c) {
				return get(x.right, key, d);
			} else if (d < key.length() - 1) {
				return get(x.mid, key, d + 1);
			} else {
				return x;
			}
		}

		public int get(final String key) {
			final Node x = get(root, key, 0);
			if (x == null) {
				return 0;
			}
			return x.val;
		}

		private Node put(Node x, final String key, final int val, final int d) {
			final char c = key.charAt(d);
			if (x == null) {
				x = new Node();
				x.c = c;
				x.val = val;
			}
			if (c < x.c) {
				x.left = put(x.left, key, val, d);
			} else if (c > x.c) {
				x.right = put(x.right, key, val, d);
			} else if (d < key.length() - 1) {
				x.mid = put(x.mid, key, val, d + 1);
			} else {
				x.val = val;
			}
			return x;
		}

		public void put(final String key, final int val) {
			root = put(root, key, val, 0);
		}

		private Node putAll(Node x, final String key, final int val, final int d) {
			final char c = key.charAt(d);
			if (x == null) {
				x = new Node();
				x.c = c;
				x.val = val;
			}
			if (c < x.c) {
				x.left = putAll(x.left, key, val, d);
			} else if (c > x.c) {
				x.right = putAll(x.right, key, val, d);
			} else if (d < key.length() - 1) {
				x.mid = putAll(x.mid, key, val, d + 1);
			}
			x.val++;
			return x;
		}

		public void putAll(final String key) {
			root = putAll(root, key, 0, 0);
		}

		private Node putAllButRoot(Node x, final String key, final int val,
				final int d) {
			final char c = key.charAt(d);
			if (x == null) {
				x = new Node();
				x.c = c;
				x.val = val;
			}
			if (c < x.c) {
				x.left = putAllButRoot(x.left, key, val, d);
			} else if (c > x.c) {
				x.right = putAllButRoot(x.right, key, val, d);
			} else if (d < key.length() - 1) {
				x.mid = putAllButRoot(x.mid, key, val, d + 1);
			}
			if (x != root) {
				x.val++;
			}
			return x;
		}

		public void putAllButRoot(final String key, final int val) {
			root = putAllButRoot(root, key, val, 0);
		}
	}

	public static void main(final String[] args) throws NumberFormatException,
	IOException {

		final TST<Integer> st = new TST<Integer>();

		final BufferedReader sc = new BufferedReader(new InputStreamReader(
				System.in));
		// BufferedReader sc = new BufferedReader(new FileReader("3.in"));

		final int n = Integer.parseInt(sc.readLine());

		final String firstWord = sc.readLine();
		final char p = firstWord.charAt(0);
		System.out.println(0);
		st.put(firstWord, 1);

		for (int i = 1; i < n; i++) {
			final String word = sc.readLine();
			if (!st.contains(word)) { // new word!
				System.out.println(st.get(word));
				if (word.charAt(0) == p) {
					st.putAll(word);
				} else {
					st.putAllButRoot(word, 0);
				}
			} else { // word exists
				System.out.println(st.get(word));
				if (word.charAt(0) == p) {
					st.putAll(word);
				} else {
					st.putAllButRoot(word, 0);
				}
			}
		}
	}
}

File: src/PiotBingTST.ts
/**
 * Ternary Search Tree (TST) implementation.
 * 
 * A more memory-efficient alternative to Tries, especially for sparse alphabets.
 * Instead of an array of child pointers, each node has only 3 children:
 * - left: characters less than the current character
 * - mid: the next character in the sequence
 * - right: characters greater than the current character
 * 
 * Benefits:
 * - Uses significantly less memory for sparse datasets
 * - Good for competitive programming constraints
 * 
 * Tradeoffs:
 * - Slightly more complex logic
 * - Similar time complexity but with larger constants
 * 
 * Time Complexity:
 *   - insert: O(m * log(alphabet_size)) on average
 *   - search: O(m * log(alphabet_size)) on average
 *   - worst case: O(m * alphabet_size)
 * Space: O(n) where n is number of nodes (much less than Trie for sparse data)
 */
export class PiotBingTST {
  private root: TSTNode | null = null;

  /**
   * Check if a word exists in the tree.
   * @param word - The word to search for
   * @returns true if the word has been inserted, false otherwise
   */
  contains(word: string): boolean {
    return this.get(word) !== 0;
  }

  /**
   * Get the frequency count of a word.
   * @param word - The word to search for
   * @returns The frequency count, or 0 if not found
   */
  get(word: string): number {
    const node = this.findNode(word, this.root, 0);
    return node === null ? 0 : node.value;
  }

  /**
   * Insert a word (overwrites existing count).
   * @param word - The word to insert
   * @param value - The frequency count to set
   */
  put(word: string, value: number): void {
    this.root = this.insertNode(this.root, word, value, 0);
  }

  /**
   * Insert a word and increment the count of all prefixes.
   * Example: inserting "cat" increments: c, ca, cat
   * @param word - The word to insert
   */
  putAll(word: string): void {
    this.root = this.incrementAll(this.root, word, 0);
  }

  /**
   * Insert a word but only increment non-root nodes.
   * Useful for avoiding double-counting the root node.
   * @param word - The word to insert
   * @param value - Not used (kept for signature compatibility)
   */
  putAllButRoot(word: string, value: number): void {
    this.root = this.incrementAllButRoot(this.root, word, value, 0);
  }

  /**
   * Recursively find a node matching the word.
   */
  private findNode(word: string, node: TSTNode | null, index: number): TSTNode | null {
    if (node === null) {
      return null;
    }

    const currentChar = word[index];

    if (currentChar < node.char) {
      return this.findNode(word, node.left, index);
    } else if (currentChar > node.char) {
      return this.findNode(word, node.right, index);
    } else if (index < word.length - 1) {
      // Move to next character in the word
      return this.findNode(word, node.mid, index + 1);
    } else {
      // Found the complete word
      return node;
    }
  }

  /**
   * Recursively insert a node, setting its value to the specified number.
   */
  private insertNode(node: TSTNode | null, word: string, value: number, index: number): TSTNode {
    const currentChar = word[index];

    if (node === null) {
      node = new TSTNode(currentChar, value);
    }

    if (currentChar < node.char) {
      node.left = this.insertNode(node.left, word, value, index);
    } else if (currentChar > node.char) {
      node.right = this.insertNode(node.right, word, value, index);
    } else if (index < word.length - 1) {
      node.mid = this.insertNode(node.mid, word, value, index + 1);
    } else {
      node.value = value;
    }

    return node;
  }

  /**
   * Recursively insert and increment counts for all prefixes.
   */
  private incrementAll(node: TSTNode | null, word: string, index: number): TSTNode {
    const currentChar = word[index];

    if (node === null) {
      node = new TSTNode(currentChar, 1);
    }

    if (currentChar < node.char) {
      node.left = this.incrementAll(node.left, word, index);
    } else if (currentChar > node.char) {
      node.right = this.incrementAll(node.right, word, index);
    } else if (index < word.length - 1) {
      node.mid = this.incrementAll(node.mid, word, index + 1);
    }

    node.value++;
    return node;
  }

  /**
   * Recursively insert and increment counts for all prefixes except root.
   */
  private incrementAllButRoot(node: TSTNode | null, word: string, value: number, index: number): TSTNode {
    const currentChar = word[index];

    if (node === null) {
      node = new TSTNode(currentChar, value);
    }

    if (currentChar < node.char) {
      node.left = this.incrementAllButRoot(node.left, word, value, index);
    } else if (currentChar > node.char) {
      node.right = this.incrementAllButRoot(node.right, word, value, index);
    } else if (index < word.length - 1) {
      node.mid = this.incrementAllButRoot(node.mid, word, value, index + 1);
    }

    // Only increment if this is not the root node
    if (node !== this.root) {
      node.value++;
    }

    return node;
  }
}

/**
 * Represents a single node in the Ternary Search Tree.
 * Each node has at most 3 children: left, mid, right.
 */
class TSTNode {
  char: string;
  value: number;
  left: TSTNode | null = null;
  mid: TSTNode | null = null;
  right: TSTNode | null = null;

  constructor(char: string, value: number = 0) {
    this.char = char;
    this.value = value;
  }
}
